Missing data-section in ggPMX controller created from nlmixr-object

Hi,

I tried to reproduce the theophyllin example using nlmixr but I have some trouble creating a working controller. The "param" and the "plot" section in the controller are created correctly but the "data" section is missing. And when I want to plot e.g. a vpc, no plot appears and NULL is given as output.

There is a warning when compiling the model and I'm not sure whether this is related to ggPMX or rather to nlmixr itself. (I have not used nlmixr for a while)

But using the theophyllin()-function, I can create a controller correctly with "param", "data" and "plot" section.

I have following code and outputs:
```
library(ggplot2)
library(nlmixr)
library(ggPMX)

one.cmt <- function() {
  ini({
    ## You may label each parameter with a comment
    tka <- 0.45 # Log Ka
    tcl <- log(c(0, 2.7, 100)) # Log Cl
    ## This works with interactive models
    ## You may also label the preceding line with label("label text")
    tv <- 3.45; label("log V")
    ## the label("Label name") works with all models
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    linCmt() ~ add(add.sd)
  })
}

fit <- nlmixr(one.cmt, theo_sd, est="saem", list(print=0))

ctr <- pmx_nlmixr(fit)
```
```
add npde
Compiling NPDE model...done
done
Compiling model...rx_a52911068df0162e01d1fa05ed7138f6_x64.c: In function 'rx_a52911068df0162e01d1fa05ed7138f6_x64_Rate':
rx_a52911068df0162e01d1fa05ed7138f6_x64.c:611:7: warning: 'rx_pred_' is used uninitialized in this function [-Wuninitialized]
   pred=rx_pred_;

done
Compiling VPC model...done
done (1.27 sec)
```


```
print(ctr)
```

```
pmx object:


|PARAM                |VALUE |
|:--------------------|:-----|
|working directory    |      |
|Modelling input file |      |
|dv                   |DV    |
|dvid                 |      |


|plot_name       |plot_type |
|:---------------|:---------|
|abs_iwres_ipred |SCATTER   |
|iwres_ipred     |SCATTER   |
|iwres_time      |SCATTER   |
|iwres_dens      |PMX_DENS  |
|iwres_qq        |PMX_QQ    |
|npde_time       |SCATTER   |
|npde_pred       |SCATTER   |
|npde_qq         |PMX_QQ    |
|dv_pred         |SCATTER   |
|dv_ipred        |SCATTER   |
|individual      |IND       |
|eta_hist        |DIS       |
|eta_box         |DIS       |
|eta_matrix      |ETA_PAIRS |
|eta_qq          |PMX_QQ    |
```
```
ctr %>% pmx_plot_vpc

NULL
```
```
ctr <- theophylline()

print(ctr)

pmx object:


|PARAM                |VALUE        |
|:--------------------|:------------|
|working directory    |theophylline |
|Modelling input file |data_pk.csv  |
|dv                   |Y            |
|dvid                 |dvid         |
|cats                 |SEX          |
|conts                |WT0,AGE0     |
|strats               |STUD         |


|data_name   |data_file       |data_label                                |
|:-----------|:---------------|:-----------------------------------------|
|predictions |predictions.txt |model predictions file                    |
|estimates   |estimates.txt   |parameter estimates file                  |
|eta         |indiv_eta.txt   |invidual estimates of random effects file |
|finegrid    |finegrid.txt    |finegrid file                             |
|input       |data_pk.csv     |modelling input                           |


|plot_name       |plot_type |
|:---------------|:---------|
|abs_iwres_ipred |SCATTER   |
|iwres_ipred     |SCATTER   |
|iwres_time      |SCATTER   |
|iwres_dens      |PMX_DENS  |
|iwres_qq        |PMX_QQ    |
|npde_time       |SCATTER   |
|npde_pred       |SCATTER   |
|npde_qq         |PMX_QQ    |
|dv_pred         |SCATTER   |
|dv_ipred        |SCATTER   |
|individual      |IND       |
|eta_hist        |DIS       |
|eta_box         |DIS       |
|eta_matrix      |ETA_PAIRS |
|eta_cats        |ETA_COV   |
|eta_conts       |ETA_COV   |
|eta_qq          |PMX_QQ    |
|pmx_vpc         |VPC       |

```

**Packages and platform**
 - Windows 10
 - R version 4.0.4
 - ggPMX version 1.2.3
 - ggplot version 3.3.4
 - nlmixr version 2.0.4
 - RxODE version 1.0.9

Do you see, where the error might be?

Many thanks in advance!

@mattfidler 

I also have a similar problem.
Interestingly, If I used the function “theophylline()” for quick reference, it is OK. I can see an excellent vpc plot. 
ctr <- theophylline()

However, if I test using the sample code listed in ggPMX-guide.pdf, I can see only “NULL” output. 
 
I used
nlmixr version = 2.0.6
ggPMX=1.2.4
R version = 4.1.2
Rstudio = 2021.09.1

YJ

``` r
library(RxODE)
library(nlmixr)
library(ggPMX)

one.cmt <- function() {
  ini({
    ## You may label each parameter with a comment
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    ## This works with interactive models
    ## You may also label the preceding line with label("label text")
    tv <- 3.45; label("log V")
    ## the label("Label name") works with all models
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    linCmt() ~ add(add.sd)
  })
}
fit <- nlmixr(one.cmt, theo_sd, est="saem", control=list(print=0))
## The fit object is a nlmixr fit; You can read it into the nlmixr controller by:
ctr <- pmx_nlmixr(fit)
ctr %>% pmx_plot_vpc()

```

output
---
```
> library(RxODE)
RxODE 1.1.2 using 4 threads (see ?getRxThreads)
  no cache: create with `rxCreateCache()`
> library(nlmixr)
> library(ggPMX)
Registered S3 method overwritten by 'GGally':
  method from   
  +.gg   ggplot2
> 
> one.cmt <- function() {
+   ini({
+     ## You may label each parameter with a comment
+     tka <- 0.45 # Log Ka
+     tcl <- 1 # Log Cl
+     ## This works with interactive models
+     ## You may also label the preceding line with label("label text")
+     tv <- 3.45; label("log V")
+     ## the label("Label name") works with all models
+     eta.ka ~ 0.6
+     eta.cl ~ 0.3
+     eta.v ~ 0.1
+     add.sd <- 0.7
+   })
+   model({
+     ka <- exp(tka + eta.ka)
+     cl <- exp(tcl + eta.cl)
+     v <- exp(tv + eta.v)
+     linCmt() ~ add(add.sd)
+   })
+ }
> fit <- nlmixr(one.cmt, theo_sd, est="saem", control=list(print=0))
i parameter labels from comments will be replaced by 'label()'
> generate SAEM model
√ done
Calculating covariance matrix
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

> creating full model...
> pruning branches (`if`/`else`)...
√ done
> loading into symengine environment...
√ done
> compiling EBE model...
√ done
Calculating residuals/tables
done
> ## The fit object is a nlmixr fit; You can read it into the nlmixr controller by:
> ctr <- pmx_nlmixr(fit)
add npde
Compiling NPDE model...done
done
Compiling model...rx_5b58d2719a7cc24ad8f4f3b24f728965_x64.c: In function 'rx_5b58d2719a7cc24ad8f4f3b24f728965_x64_Rate':
rx_5b58d2719a7cc24ad8f4f3b24f728965_x64.c:611:7: warning: 'rx_pred_' is used uninitialized in this function [-Wuninitialized]
   pred=rx_pred_;
   ~~~~^~~~~~~~~

done
Compiling VPC model...done
done (2.24 sec)
>                   
> ctr %>% pmx_plot_vpc()
NULL
> 
> 
> ```
Error: attempt to use zero-length variable name
> 
```
---

--- a/R/nlmixr.R
+++ b/R/nlmixr.R
@@ -15,7 +15,7 @@
 
 pmx_nlmixr <- function(fit, dvid, conts, cats, strats, endpoint, settings, vpc = TRUE) {
   EFFECT <- EVID <- ID <- MDV <- NULL
-  
+
   if (missing(fit)) {
     return(NULL)
   }

--- a/R/pmxClass.R
+++ b/R/pmxClass.R
@@ -27,9 +27,6 @@ pmx_sim <- function(
     }
     id_col <- grep("^id$", names(sim), ignore.case = TRUE, value = TRUE)
     setnames(sim, id_col, "ID")
-    if (!inherits(sim$ID, "factor")) {
-      sim[, ID := as.integer(ID)]
-    }
     obj <- list(
       sim = sim,
       idv = idv,
@@ -168,7 +165,7 @@ pmx_mlxtran <- function(file_name, config = "standing", call = FALSE, endpoint, 
   if (grepl("*", file_name, fixed = TRUE)) {
     assert_that(version>=0, msg = "Using wildcard in file_name assume providing non-negative version")
     file_name <- gsub("*", version, file_name, fixed = TRUE)
-  } 
+  }
   params <- parse_mlxtran(file_name)
   rr <- as.list(match.call()[-1])
   rr$file_name <- NULL
@@ -189,7 +186,7 @@ pmx_mlxtran <- function(file_name, config = "standing", call = FALSE, endpoint, 
   params$call <- NULL
   # We don't need to pass version to pmx_mlx
   params$version <- NULL
-  
+
   do.call(pmx_mlx, params)
 }
 
@@ -939,13 +936,7 @@ pmx_initialize <- function(self, private, data_path, input, dv,
         call. = FALSE
       )
     }
-    if (inherits(dx$ID,"factor") & !inherits(inn$ID,"factor")) {
-      inn[, ID := factor(ID, levels = levels(ID))]
-    }
-    if (!inherits(dx$ID, "factor") & inherits(inn$ID, "factor")) {
-      dx[, ID := factor(ID, levels = levels(ID))]
-    }
-    self$data[["sim"]] <- merge(dx, inn, by = c("ID", "TIME"))
+    self$data[["sim"]] <- merge_dx_inn_by_id_time(dx, inn, config$sys)
     self$sim <- sim
   }
 

--- a/R/utils.R
+++ b/R/utils.R
@@ -299,11 +299,11 @@ parse_mlxtran <- function(file_name) {
   conts <- dat[grepl("use=covariate, type=continuous", value), key]
   ## occ
   occ <- dat[grepl("use=occasion", value), key]
-  ## id 
+  ## id
   id <- dat[grepl("use=identifier", value), key]
-  ## time 
+  ## time
   time <- dat[grepl("use=time", value), key]
-  
+
 
 
   res <- list(
@@ -379,3 +379,22 @@ pk_pd <- function(code = "3") {
 
 
 quantile <- function(...) do.call(stats::quantile, list(...))
+
+#' Merge dx into inn, making sure that the IDs match types
+#' @param dx input data.table
+#' @param inn merging to data.table
+#' @param sys System to consider; Currently will change integer simulations in nlmixr to factors
+#' @return merged data.table
+#' @author Matthew L. Fidler
+#' @noRd
+merge_dx_inn_by_id_time <- function(dx, inn, sys) {
+  if (sys == "nlmixr") {
+    if (inherits(inn$ID, "factor")) {
+      ID <- dx$ID
+      attr(ID, "levels") <- levels(inn$ID)
+      attr(ID, "class") <- "factor"
+      dx$ID <- ID
+    }
+  }
+  merge(dx, inn, by = c("ID", "TIME"))
+}

